### DISCUSSION TITLE
Offer to set up ESLint in create-astro

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [withastro/roadmap#621] [Proposal]
## Summary

ESLint is generally considered to be a _useful_, even _necessary_, tool for JavaScript/TypeScript web apps. However, configuring it can involve wrangling several configurations and plugins together. It would be useful for many Astro users if `npm create astro` et al. offered to set up a starting ESLint configuration.

## Background & Motivation

<details>
<summary>Contained here are the most common setup points I think many Astro would need.</summary>

* _[Required]_ ["Stock"/"vanilla" ESLint](https://eslint.org/docs/latest/use/getting-started#manual-set-up): an `.eslintrc.cjs` with `env`, `extends`, and `parserOptions`
* _[Required]_ [`eslint-plugin-astro`](https://github.com/ota-meshi/eslint-plugin-astro): `extends` and `overrides`
* _[Required for TypeScript]_ [Adding typescript-eslint](https://typescript-eslint.io/getting-started): `extends`, `parser`, and `plugins`
    * Note that this requires modifying the `overrides` configuration for `eslint-plugin-astro`
    * Note also that the [`@typescript-eslint/triple-slash-reference` rule](https://typescript-eslint.io/rules/triple-slash-reference) fires on `src/env.d.ts`, so you'll need to turn it off for that file or use an `.eslintignore`...
* _[Optional, must be using TypeScript]_ [Adding typescript-eslint type information](https://typescript-eslint.io/linting/typed-linting): `extends` and `parserOptions`
* _[Optional, but almost everyone should use it]_ [`eslint-plugin-jsx-a11y`](https://npmjs.com/package/eslint-plugin-jsx-a11y): `extends` and `plugins`
* _[Optional]_ Adding framework-specific plugins, e.g.:
    * React: [`eslint-plugin-react`](https://www.npmjs.com/package/eslint-plugin-react) and [`eslint-plugin-react-hooks`](https://www.npmjs.com/package/eslint-plugin-react-hooks): `extends` and `plugins`
    * Solid: [`eslint-plugin-solid`](https://www.npmjs.com/package/eslint-plugin-solid): `extends` and `plugins` 
</details>

You can see an example of an ESLint config for Astro with Solid here: https://github.com/JoshuaKGoldberg/joshuakgoldberg-dot-com/blob/7649e1fa12280ffdbf16c9887e8724de0b3203ab/.eslintrc.cjs. 

Note that many of these plugins require using [ESLint `overrides`](https://eslint.org/docs/latest/use/configure/configuration-files#configuration-based-on-glob-patterns) in specific ways. For example, if you apply the Solid plugin to `.astro` files, its [`solid/prefer-for` rule](https://github.com/solidjs-community/eslint-plugin-solid/blob/0e037ef35e6c2e0538eaacd57efcb63e25f332c9/docs/prefer-for.md) will rewrite your Astro JSX loops to try to use Solid's `<For>`. I learned that the hard way 😬 and it was very confusing+inconvenient at first.

That's a lot of configuration - and the vast majority of users aren't going to be able to get there. I'm a maintainer on typescript-eslint and it took me a few tries!

### Third Party Integrations

Each of the framework plugins -including Astro's- that would be installed are third-party plugins not maintained by Astro. This RFC's proposal effectively adds a dependency on those external plugins being trustworthy.

These plugins are already key parts of the Astro ecosystem: especially [`astro-eslint-parser`](https://npmjs.com/package/astro-eslint-parser) and [`eslint-plugin-astro`](https://ota-meshi.github.io/eslint-plugin-astro). An [estimated >80% of survey-taking JS developers use ESLint](https://2020.stateofjs.com/en-US/other-tools/#utilities). https://docs.astro.build/en/editor-setup/#eslint already points towards `eslint-plugin-astro`. We can assume users are already treating these plugins as a core part of the Astro experience. Directing more users to the plugins will help raise their visibility.

## Goals

- Automate the process of setting up a _"good"_ starter ESLint configuration for users based on their project needs
- Improve Astro app development flow & reliability by providing more comprehensive linting
- Increase community usage of specifically the Astro ESLint parser and plugin

## Example

How about in the Houston CLI experience, it also asks if we want an ESLint config?

```plaintext
...

  deps   Install dependencies?
         Yes
      ✔  Dependencies installed

    ts   Do you plan to write TypeScript?
         Yes

   use   How strict should TypeScript be?
         Strict
      ✔  TypeScript customized

  lint   Would you like to add an ESLint configuration?
         Yes

config   What ESLint plugin(s) would you like to add?
         ● astro       (recommended)
         ● jsx-a11y    (recommended)
         ● typescript  (recommended)
         ○ alpinejs
         ○ lit
         ○ preact
         ○ react
         ○ react-hooks
         ○ solid-js
         ○ svelte
         ○ vue
...
```

...and if they're using a template with a view library, it should be included?

```plaintext
...
         ● typescript  (recommended)
         ● react       (recommended for your template)
         ● react-hooks (recommended for your template)     
         ○ alpinejs         
...
```

This would require installing dependencies later in the setup script.

## Acknowledgements

Thanks to @ElianCodes for feedback on drafts of this RFC :heart:

**Comment by bluwy** [comment]
I think there was discussion to use `astro add` instead, so it won't be a decision only available on creation. However, currently `astro add` works with integrations only.

**Reply by JoshuaKGoldberg** [reply]
Ah, right! This rings a bell. I do agree that it would be nice to have there _and_ on creation.

**Comment by vitch** [comment]
I found this while struggling to get a nice setup with the current eslint and all the necessary plugins in astro. Has there been any change in the status quo? I think the RFC was written before eslint flat config was a thing?

I'm new to astro but not to js/ts/eslint etc. My first thought was for an `astro add eslint` command which set stuff up although I don't yet know what the limitations of integrations are... I'll see if I can get an example of a straightforward setup and then if there is interest I could take a look at making it repeatable.
